In [1]:
!pip install neuralforecast

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 170.6/170.6 kB 2.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 776.9/776.9 kB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.6/62.6 MB 10.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 413.4/413.4 kB 17.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 806.1/806.1 kB 33.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 101.7/101.7 kB 9.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.6/230.6 kB 12.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 5.7 MB/s eta 0:00:00


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import math
import numpy as np
import torch
from typing import Optional, Union, Tuple


import torch.nn.functional as F
from torch.distributions import Distribution
from torch.distributions import Bernoulli, Normal, StudentT, Poisson, NegativeBinomial

from torch.distributions import constraints


from neuralforecast import NeuralForecast
from neuralforecast.models import Informer
from neuralforecast.tsdataset import TimeSeriesDataset
from neuralforecast.utils import AirPassengers, AirPassengersPanel, AirPassengersStatic
from neuralforecast.losses.numpy import rmse, mape
from pandas.tseries.holiday import USFederalHolidayCalendar
from pandas.tseries.offsets import CustomBusinessDay
from neuralforecast.losses.pytorch import MQLoss, DistributionLoss, MSE, MAE

In [3]:
from numpy.random import seed
from random import randrange

In [4]:
df = pd.read_excel("SP100_vol (clean).xlsx", index_col="Date",parse_dates=True)
columns=df.columns
df_static=pd.read_excel("Static_df.xlsx")
n_series=len(df.columns)
train = df.iloc[:int(len(df['NKE'])*0.7)]
validation_length=int(len(df['NKE'])*0.7)-int(len(df['NKE'])*0.5)
train=train.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

In [5]:
n_inputs = [5,10,15,21,42,84]
hidden_size = [60,90,120,150,180]
epochs=[50,100,150,250,350,450,550,650,750]
learning_rate=[0.0005,0.0001,0.00005,0.00001]
num_lr_decays=[5,3,2,1]
scaler_type=["robust","standard",'minmax']
losses=[MSE(),MAE(),MQLoss(level=[80, 90]),DistributionLoss(distribution='StudentT', level=[80, 90])]
n_head=[2,3,4,5,6]
decoder_input_size_multiplier=[0.3,0.4,0.5,0.6,0.7]
dropout=[0,0.2,0.3,0.4]
factor=[2,3,4]
conv_hidden_size=[12,22,32,42,52]
decoder_layers = [1,2,3]

# Hyperparameter Search

In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 58088380


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 8265540


Epoch: 750 | Inputs: 15 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 2 | decoder_layers: 3
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 5
RMSE: 0.002623957527121715 | QLIKE: 0.024472820072106365 | MAE: 0.001594304470502678 | MAPE: 0.15745243874819348


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 94365581


Epoch: 550 | Inputs: 84 | hidden_size: 60 | Loss: DistributionLoss() | Scaler Type: robust | factor: 3 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.6 | Dropout: 0.4 | n_head: 4
RMSE: 0.0032222245158293954 | QLIKE: 0.0388421359900238 | MAE: 0.0019946754199625335 | MAPE: 0.18399091292022482


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 62810548


Epoch: 150 | Inputs: 10 | hidden_size: 90 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.6 | Dropout: 0.4 | n_head: 3
RMSE: 0.004168148094935638 | QLIKE: 0.05031963011862237 | MAE: 0.002895360174243847 | MAPE: 0.31730757132913734


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 116843341


Epoch: 650 | Inputs: 15 | hidden_size: 150 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.5 | Dropout: 0.4 | n_head: 6
RMSE: 0.0026472110803416817 | QLIKE: 0.02637026577618137 | MAE: 0.0015996838910256558 | MAPE: 0.1582931872702211


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 3 | hidden_size: 60 | Loss: MAE() | Scaler Type: robust | factor: 3 | decoder_layers: 2
Learning Rate: 2 | decoder_input_size_multiplier: 0.3 | Dropout: 0.2 | n_head: 4
RMSE: 0.0026923646923624396 | QLIKE: 0.02562070503979714 | MAE: 0.0015801182520588744 | MAPE: 0.1527446520741483


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 99692143


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 41131853


Epoch: 50 | Inputs: 42 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 4 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 4 | encoder_layers: 4
RMSE: 0.0031401663990131595 | QLIKE: 0.036772443671046304 | MAE: 0.001992331963106507 | MAPE: 0.19306762540718506


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 53427168


Epoch: 50 | Inputs: 10 | hidden_size: 150 | Loss: MSE() | Scaler Type: minmax | factor: 4 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.6 | Dropout: 0 | n_head: 3 | encoder_layers: 2
RMSE: 0.002672827436627847 | QLIKE: 0.025216928783982014 | MAE: 0.0016547970641475854 | MAPE: 0.16369663654092323


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 51869160


Epoch: 100 | Inputs: 21 | hidden_size: 180 | Loss: MSE() | Scaler Type: robust | factor: 4 | decoder_layers: 1
Learning Rate: 3 | decoder_input_size_multiplier: 0.5 | Dropout: 0 | n_head: 2 | encoder_layers: 1
RMSE: 0.002871570627432254 | QLIKE: 0.02872342736051522 | MAE: 0.0017220385114938596 | MAPE: 0.16854298280295113


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 52829811


Epoch: 650 | Inputs: 15 | hidden_size: 180 | Loss: MAE() | Scaler Type: standard | factor: 2 | decoder_layers: 1
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 3 | encoder_layers: 4
RMSE: 0.002607734968576004 | QLIKE: 0.02455007467345509 | MAE: 0.0015312307798305714 | MAPE: 0.14579977285890083


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 550 | Inputs: 15 | hidden_size: 90 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 6 | encoder_layers: 1
RMSE: 0.002555410868170968 | QLIKE: 0.02348263351971576 | MAE: 0.0014879118305409647 | MAPE: 0.14372182774158154


In [ ]:
for z in range (5):
  i=randrange(7)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 37525783


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 68528278


Epoch: 650 | Inputs: 3 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: minmax | factor: 4 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 2 | encoder_layers: 3
RMSE: 0.002616405669726207 | QLIKE: 0.023482675208007297 | MAE: 0.0015623582165877623 | MAPE: 0.15566522091518595


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 3933120


Epoch: 550 | Inputs: 84 | hidden_size: 120 | Loss: MSE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 3 | decoder_input_size_multiplier: 0.7 | Dropout: 0 | n_head: 5 | encoder_layers: 2
RMSE: 0.0029853797130454064 | QLIKE: 0.03131212999467374 | MAE: 0.0019278649133770206 | MAPE: 0.19244788434737384


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 39520493


Epoch: 450 | Inputs: 3 | hidden_size: 120 | Loss: MAE() | Scaler Type: robust | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 5 | encoder_layers: 2
RMSE: 0.0026133499558680722 | QLIKE: 0.024442106249439217 | MAE: 0.0015080748243778562 | MAPE: 0.14455508564684016


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 100595084


Epoch: 650 | Inputs: 5 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: robust | factor: 3 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 4 | encoder_layers: 3
RMSE: 0.002652942202126921 | QLIKE: 0.0247917757098761 | MAE: 0.0015481489741161947 | MAPE: 0.1483219500255115


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 750 | Inputs: 84 | hidden_size: 60 | Loss: MQLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 6 | encoder_layers: 3
RMSE: 0.003627676931846737 | QLIKE: 0.04991559403697488 | MAE: 0.002337886097000056 | MAPE: 0.22782533845173578


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 64785404


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 74056557


Epoch: 50 | Inputs: 21 | hidden_size: 90 | Loss: DistributionLoss() | Scaler Type: robust | factor: 2 | decoder_layers: 1
Learning Rate: 1 | decoder_input_size_multiplier: 0.6 | Dropout: 0.4 | n_head: 2 | encoder_layers: 4
RMSE: 0.0030158701852697908 | QLIKE: 0.03173590582738745 | MAE: 0.0018665859359784645 | MAPE: 0.18147310206239328


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 76370125


Epoch: 50 | Inputs: 3 | hidden_size: 90 | Loss: MQLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 5 | encoder_layers: 4
RMSE: 0.002887363165324041 | QLIKE: 0.027388386723191712 | MAE: 0.0017862439510503679 | MAPE: 0.1805360994048373


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 40897195


Epoch: 50 | Inputs: 42 | hidden_size: 120 | Loss: MAE() | Scaler Type: standard | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 2 | encoder_layers: 4
RMSE: 0.002895539240472086 | QLIKE: 0.03052724485984949 | MAE: 0.001775814746647289 | MAPE: 0.16834944135048738


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 80010322


Epoch: 50 | Inputs: 3 | hidden_size: 60 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 5 | encoder_layers: 3
RMSE: 0.002574688332343211 | QLIKE: 0.02464630610807486 | MAE: 0.0014600210075708614 | MAPE: 0.13913911479760718


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 350 | Inputs: 15 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: minmax | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 2 | encoder_layers: 4
RMSE: 0.0032106023221640145 | QLIKE: 0.026319196347376797 | MAE: 0.002253339316871675 | MAPE: 0.24286359049663542


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 7901981


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 111022087


Epoch: 100 | Inputs: 21 | hidden_size: 180 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 4 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.7 | Dropout: 0.4 | n_head: 2 | encoder_layers: 3
RMSE: 0.002742590869821286 | QLIKE: 0.028425066109220288 | MAE: 0.0016869412217203309 | MAPE: 0.16255579791165312


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 5129103


Epoch: 550 | Inputs: 21 | hidden_size: 60 | Loss: MAE() | Scaler Type: minmax | factor: 4 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.3 | Dropout: 0.2 | n_head: 4 | encoder_layers: 2
RMSE: 0.0026064945192532467 | QLIKE: 0.024429020828485616 | MAE: 0.0015334284254808356 | MAPE: 0.1457647500987668


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 34845653


Epoch: 350 | Inputs: 21 | hidden_size: 120 | Loss: DistributionLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 1 | decoder_input_size_multiplier: 0.3 | Dropout: 0.4 | n_head: 3 | encoder_layers: 1
RMSE: 0.00269626716780776 | QLIKE: 0.028367632339246607 | MAE: 0.0016125197987794525 | MAPE: 0.15416511497580263


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 42946116


Epoch: 550 | Inputs: 5 | hidden_size: 120 | Loss: MSE() | Scaler Type: minmax | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 4 | encoder_layers: 2
RMSE: 0.00266059444879744 | QLIKE: 0.02438844142957604 | MAE: 0.0016346592884275346 | MAPE: 0.16459980073777064


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 650 | Inputs: 21 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: robust | factor: 3 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.6 | Dropout: 0.3 | n_head: 5 | encoder_layers: 3
RMSE: 0.0026917350837004465 | QLIKE: 0.025490133730841417 | MAE: 0.0015636787452678468 | MAPE: 0.14852668667703187


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 99911194


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 11790293


Epoch: 550 | Inputs: 3 | hidden_size: 120 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.3 | Dropout: 0.2 | n_head: 3 | encoder_layers: 4
RMSE: 0.002630317785762633 | QLIKE: 0.024600802193315643 | MAE: 0.0015220216545571358 | MAPE: 0.1463991353323181


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 80160178


Epoch: 350 | Inputs: 3 | hidden_size: 120 | Loss: MAE() | Scaler Type: minmax | factor: 4 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 2 | encoder_layers: 2
RMSE: 0.0025863206439261143 | QLIKE: 0.023604445853552498 | MAE: 0.0015205083952160093 | MAPE: 0.14836724247936472


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 88088879


Epoch: 50 | Inputs: 15 | hidden_size: 60 | Loss: MAE() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0.2 | n_head: 4 | encoder_layers: 2
RMSE: 0.0032274940156912266 | QLIKE: 0.03709517547210229 | MAE: 0.0020408560577903417 | MAPE: 0.2000448351893014


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 14803130


Epoch: 350 | Inputs: 5 | hidden_size: 150 | Loss: MQLoss() | Scaler Type: standard | factor: 4 | decoder_layers: 1
Learning Rate: 3 | decoder_input_size_multiplier: 0.5 | Dropout: 0.3 | n_head: 6 | encoder_layers: 4
RMSE: 0.0026808966262370786 | QLIKE: 0.02535409006808485 | MAE: 0.0016112814535082353 | MAPE: 0.15413182553863147


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 50 | Inputs: 3 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: robust | factor: 2 | decoder_layers: 2
Learning Rate: 1 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 6 | encoder_layers: 3
RMSE: 0.0027231367573523624 | QLIKE: 0.026166193285566935 | MAE: 0.0015726901108348863 | MAPE: 0.15295641364955564


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 84661054


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 56717223


Epoch: 550 | Inputs: 15 | hidden_size: 180 | Loss: MAE() | Scaler Type: robust | factor: 3 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.7 | Dropout: 0.2 | n_head: 2 | encoder_layers: 1
RMSE: 0.0026610215620732573 | QLIKE: 0.02501290257281625 | MAE: 0.001571435770276844 | MAPE: 0.14960839780472046


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 932969


Epoch: 150 | Inputs: 5 | hidden_size: 60 | Loss: MQLoss() | Scaler Type: robust | factor: 3 | decoder_layers: 1
Learning Rate: 3 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 4 | encoder_layers: 1
RMSE: 0.003152776323159092 | QLIKE: 0.03413515354346437 | MAE: 0.0019088300381322396 | MAPE: 0.18605567828408978


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 97867483


Epoch: 350 | Inputs: 5 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 2
Learning Rate: 0 | decoder_input_size_multiplier: 0.3 | Dropout: 0.3 | n_head: 4 | encoder_layers: 3
RMSE: 0.0025933894533778958 | QLIKE: 0.023898637905809792 | MAE: 0.0015274344787599603 | MAPE: 0.1466575520514694


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 124392084


Epoch: 450 | Inputs: 5 | hidden_size: 60 | Loss: DistributionLoss() | Scaler Type: standard | factor: 2 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.5 | Dropout: 0.4 | n_head: 5 | encoder_layers: 2
RMSE: 0.0025779943994150866 | QLIKE: 0.023943390949488143 | MAE: 0.0015080160782626304 | MAPE: 0.14438084834293424


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 650 | Inputs: 21 | hidden_size: 180 | Loss: MQLoss() | Scaler Type: robust | factor: 4 | decoder_layers: 3
Learning Rate: 3 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 4 | encoder_layers: 3
RMSE: 0.0028556931945265484 | QLIKE: 0.02980722723669842 | MAE: 0.001670739529996997 | MAPE: 0.15718944124070136


In [ ]:
for z in range (5):
  i=randrange(6)
  h=randrange(5)
  m=randrange(4)
  a=randrange(5)
  k=randrange(9)
  l=randrange(4)
  n=randrange(4)
  o=randrange(3)
  a1=randrange(5)
  a2=randrange(3)
  a3=randrange(5)
  a4=randrange(4)
  a5=randrange(3)
  model = Informer(h=1, input_size=n_inputs[i],
           decoder_input_size_multiplier=decoder_input_size_multiplier[a1], hidden_size=hidden_size[h],
           dropout=dropout[m], factor=factor[a2], n_head=n_head[a],
           conv_hidden_size=conv_hidden_size[a3],
           decoder_layers=decoder_layers[a5],
           loss=losses[m], max_steps=epochs[k],
           learning_rate=learning_rate[n], num_lr_decays=num_lr_decays[n],
           val_check_steps=500,
           batch_size=32,
           scaler_type=scaler_type[o], random_seed=randrange(129228148))
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Epoch: {epochs[k]} | Inputs: {n_inputs[i]} | hidden_size: {hidden_size[h]} | Loss: {losses[m]} | Scaler Type: {scaler_type[o]} | factor: {factor[a2]} | decoder_layers: {decoder_layers[a5]}")
  print(f"Learning Rate: {n} | decoder_input_size_multiplier: {decoder_input_size_multiplier[a1]} | Dropout: {dropout[m]} | n_head: {n_head[a]} | encoder_layers: {encoder_layers[a4]}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 53508068


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 12621714


Epoch: 450 | Inputs: 3 | hidden_size: 120 | Loss: DistributionLoss() | Scaler Type: minmax | factor: 4 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.6 | Dropout: 0.4 | n_head: 2 | encoder_layers: 4
RMSE: 0.0026184564756173722 | QLIKE: 0.02361816102926254 | MAE: 0.0015499775860708823 | MAPE: 0.15498521371949311


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 43452637


Epoch: 650 | Inputs: 3 | hidden_size: 60 | Loss: MSE() | Scaler Type: minmax | factor: 3 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.4 | Dropout: 0 | n_head: 3 | encoder_layers: 2
RMSE: 0.0027385232565717795 | QLIKE: 0.02476675451494178 | MAE: 0.0016680039851449777 | MAPE: 0.17053385061951215


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 21386484


Epoch: 250 | Inputs: 3 | hidden_size: 120 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.5 | Dropout: 0.4 | n_head: 3 | encoder_layers: 4
RMSE: 0.002578499101513351 | QLIKE: 0.023808318405754626 | MAE: 0.0015018842308242315 | MAPE: 0.14497347329854648


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 61014095


Epoch: 450 | Inputs: 21 | hidden_size: 120 | Loss: MQLoss() | Scaler Type: standard | factor: 4 | decoder_layers: 3
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.3 | n_head: 4 | encoder_layers: 1
RMSE: 0.0025705626558521293 | QLIKE: 0.023540719850907478 | MAE: 0.0015023156994787158 | MAPE: 0.14506989182727034


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Epoch: 150 | Inputs: 3 | hidden_size: 180 | Loss: MAE() | Scaler Type: standard | factor: 2 | decoder_layers: 3
Learning Rate: 2 | decoder_input_size_multiplier: 0.6 | Dropout: 0.2 | n_head: 6 | encoder_layers: 3
RMSE: 0.0025671840600963274 | QLIKE: 0.023870203355431007 | MAE: 0.0014796783554983245 | MAPE: 0.14253602004293936


INFO:lightning_fabric.utilities.seed:Seed set to 52829811

Epoch: 550 | Inputs: 15 | hidden_size: 90 | Loss: DistributionLoss() | Scaler Type: standard | factor: 3 | decoder_layers: 1
Learning Rate: 0 | decoder_input_size_multiplier: 0.4 | Dropout: 0.4 | n_head: 6 | encoder_layers: 1
RMSE: 0.002555410868170968 | QLIKE: 0.02348263351971576 | MAE: 0.0014879118305409647 | MAPE: 0.14372182774158154

In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Informer(h=1, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 103984860


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 65186079


Random Seed: 103984860
RMSE: 0.0025637922380353432 | QLIKE: 0.023656446427940982 | MAE: 0.0014902631317809752 | MAPE: 0.14322438348361605


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 72588291


Random Seed: 65186079
RMSE: 0.0025589563090613315 | QLIKE: 0.023911584704346074 | MAE: 0.0014809849056579582 | MAPE: 0.14190339100931462


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 20261850


Random Seed: 72588291
RMSE: 0.0025608204130160854 | QLIKE: 0.023644353131577818 | MAE: 0.0014944484962180907 | MAPE: 0.14424215456519565


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 53629349


Random Seed: 20261850
RMSE: 0.0025550934589202694 | QLIKE: 0.023552155935428466 | MAE: 0.001486502789875348 | MAPE: 0.1435164527334639


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 53629349
RMSE: 0.0025598325178645322 | QLIKE: 0.02349754649858009 | MAE: 0.0014888334688729825 | MAPE: 0.14414679467578126


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Informer(h=1, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 44695605


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 10728490


Random Seed: 44695605
RMSE: 0.0025595383530491035 | QLIKE: 0.023481209843081066 | MAE: 0.0014945417677182276 | MAPE: 0.1448448291437252


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 128364941


Random Seed: 10728490
RMSE: 0.0025531129703965333 | QLIKE: 0.023555355754186033 | MAE: 0.0014844741125339045 | MAPE: 0.14371147209178825


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 32036319


Random Seed: 128364941
RMSE: 0.0025553878101550543 | QLIKE: 0.023395640781362387 | MAE: 0.0014991082993314067 | MAPE: 0.14518306390315988


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 59706950


Random Seed: 32036319
RMSE: 0.0025470022277433 | QLIKE: 0.02326355054119016 | MAE: 0.0014856217599719332 | MAPE: 0.1443688390823151


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 59706950
RMSE: 0.002560726021228924 | QLIKE: 0.02373156801313679 | MAE: 0.0014946974172928345 | MAPE: 0.1439976390391517


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Informer(h=1, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 66767543


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 81740443


Random Seed: 66767543
RMSE: 0.0025643885292836893 | QLIKE: 0.023778920197481863 | MAE: 0.0014906881125199905 | MAPE: 0.143077488781003


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 46814803


Random Seed: 81740443
RMSE: 0.0025561397057476507 | QLIKE: 0.02343398657045093 | MAE: 0.0014931759571353584 | MAPE: 0.14449075817310508


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 69928218


Random Seed: 46814803
RMSE: 0.0025532729563544224 | QLIKE: 0.023502419362890332 | MAE: 0.0014814941562813566 | MAPE: 0.14329758844511156


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 62771770


Random Seed: 69928218
RMSE: 0.0025631638817121636 | QLIKE: 0.02382380773882715 | MAE: 0.0014885037651513374 | MAPE: 0.1430850875305739


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 62771770
RMSE: 0.00255124626599572 | QLIKE: 0.02316486949913828 | MAE: 0.0014987311861506406 | MAPE: 0.14626835054351725


In [ ]:
for z in range (5):
  randomseed=randrange(129228148)
  model = Informer(h=1, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
                random_seed=randomseed)
  fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
  forecasts = fcst.cross_validation(df=train,val_size=1,static_df=df_static,n_windows=None, test_size=validation_length-validation_length%1,step_size=1)
  forecasts = forecasts.dropna()
  if "Informer-median" not in list(forecasts.columns.values):
    Y_hat=forecasts["Informer"].values
  else:
    Y_hat=forecasts["Informer-median"].values
  Y_true=forecasts["y"].values
  RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
  QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
  MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
  MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
  print(f"Random Seed: {randomseed}")
  print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 64877732


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 91659253


Random Seed: 64877732
RMSE: 0.002554287934634896 | QLIKE: 0.023251820873380697 | MAE: 0.0014921916935440134 | MAPE: 0.14545500455632238


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 85539998


Random Seed: 91659253
RMSE: 0.0025424074049450016 | QLIKE: 0.023305209020703193 | MAE: 0.0014827350915403916 | MAPE: 0.14378904370369333


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 114170943


Random Seed: 85539998
RMSE: 0.002553751200702018 | QLIKE: 0.023382832320866957 | MAE: 0.001494081853112064 | MAPE: 0.1447098656749264


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

INFO:lightning_fabric.utilities.seed:Seed set to 50240754


Random Seed: 114170943
RMSE: 0.0025540566361160273 | QLIKE: 0.023307466569709168 | MAE: 0.001499382415859975 | MAPE: 0.1458027375241016


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

Random Seed: 50240754
RMSE: 0.0025647927281366102 | QLIKE: 0.023961074130152558 | MAE: 0.0014923820118654195 | MAPE: 0.14372477364011754


# Test

In [6]:
test_length=int(len(df['NKE'])*0.3)
df1=df.unstack().reset_index(name="Value").rename(columns={"level_0":"unique_id", "Date":"ds", "Value":"y"})

# H=1

In [ ]:
model = Informer(h=1, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
              random_seed=91659253)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=1,static_df=df_static,n_windows=None, test_size=test_length-test_length%1,step_size=1)
forecasts = forecasts.dropna()
if "Informer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Informer"].values
else:
  Y_hat=forecasts["Informer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

INFO:lightning_fabric.utilities.seed:Seed set to 91659253


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.0038727352169593337 | QLIKE: 0.03171137741753872 | MAE: 0.0024056815164261052 | MAPE: 0.17953156767113432


In [ ]:
if "Informer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Informer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

In [ ]:
df2.to_excel("Informer forecasts.xlsx")

# H=5

In [7]:
H=5

model = Informer(h=H, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
              random_seed=91659253)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "Informer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Informer"].values
else:
  Y_hat=forecasts["Informer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "Informer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Informer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("Informer forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 91659253


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.004810647503862946 | QLIKE: 0.04621980546512753 | MAE: 0.0028530351156349065 | MAPE: 0.2075608985109403


# H=10

In [8]:
H=10

model = Informer(h=H, input_size=15,
           decoder_input_size_multiplier=0.4, hidden_size=90,
           dropout=0.4, factor=3, n_head=6,
           conv_hidden_size=32,
           decoder_layers=3,
           loss=losses[3], max_steps=550,
           learning_rate=learning_rate[0], num_lr_decays=num_lr_decays[0],
           val_check_steps=500,
           batch_size=32,
           scaler_type='standard',
              random_seed=91659253)
fcst = NeuralForecast(models=[model],freq=CustomBusinessDay(calendar=USFederalHolidayCalendar()))
forecasts = fcst.cross_validation(df=df1,val_size=H,static_df=df_static,n_windows=None, test_size=test_length-test_length%H,step_size=H)
forecasts = forecasts.dropna()
if "Informer-median" not in list(forecasts.columns.values):
  Y_hat=forecasts["Informer"].values
else:
  Y_hat=forecasts["Informer-median"].values
Y_true=forecasts["y"].values
RMSE=np.sqrt(np.sum(((Y_true-Y_hat)**2))/len(Y_true))
QLIKE=(np.sum(Y_true/Y_hat-np.log(abs(Y_true)/abs(Y_hat))-1)/len(Y_true))
MAE=np.sum(abs(Y_true-Y_hat))/len(Y_true)
MAPE=np.sum(abs(Y_true-Y_hat)/(Y_true))/len(Y_true)
print(f"RMSE: {RMSE} | QLIKE: {QLIKE} | MAE: {MAE} | MAPE: {MAPE}")

if "Informer-median" not in list(forecasts.columns.values):
  df2=forecasts.drop(columns=["cutoff", "y"]).set_index("ds")
  df2=df2.pivot(columns="unique_id")
else:
  df2=forecasts[["ds","unique_id","Informer-median"]].set_index("ds")
  df2=df2.pivot(columns="unique_id")

df2.to_excel("Informer forecasts.xlsx")

INFO:lightning_fabric.utilities.seed:Seed set to 91659253


Sanity Checking: |          | 0/? [00:00<?, ?it/s]

Training: |          | 0/? [00:00<?, ?it/s]

Validation: |          | 0/? [00:00<?, ?it/s]

Predicting: |          | 0/? [00:00<?, ?it/s]

RMSE: 0.005677534183098615 | QLIKE: 0.059876085667915614 | MAE: 0.0032839211901392867 | MAPE: 0.23310020083417615
